crates/common/src/proofs.rs

```diff
@@ -92,21 +94,23 @@ impl TryFrom<SignedConstraints> for SignedConstraintsWithProofData {
     type Error = ProofError;
 
     fn try_from(value: SignedConstraints) -> Result<Self, ProofError> {
-        let transactions = value
-            .message
-            .transactions
-            .iter()
-            .map(|tx| {
-                let tx_hash = TxHash::from_slice(keccak256(tx.to_vec()).as_slice());
-                let tx_root = Transaction::try_from(tx.to_vec().as_ref())
-                    .map_err(|_| ProofError::VerificationFailed)?
-                    .hash_tree_root()
-                    .map_err(|_| ProofError::VerificationFailed)?;
-                let tx_root = Hash256::from_slice(&tx_root.to_vec());
-
-                Ok((tx_hash, tx_root))
-            })
-            .collect::<Result<Vec<_>, ProofError>>()?;
+        let mut transactions = Vec::with_capacity(value.message.transactions.len());
+        for transaction in value.message.transactions.to_vec().iter() {
+            let tx = PooledTransactionsElement::decode_enveloped(transaction.to_vec().into())
+                .map_err(|_| ProofError::DecodingFailed)?;
+
+            let tx_hash = *tx.hash();
+
+            // Compute the hash tree root on the transaction object decoded without the optional
+            // sidecar. this is to prevent hashing the blobs of type 3 transactions.
+            let root = tx.into_transaction().envelope_encoded();
+            let root =
+                Transaction::try_from(root.as_ref()).map_err(|_| ProofError::DecodingFailed)?;
+            let root = root.clone().hash_tree_root().map_err(|_| ProofError::DecodingFailed)?;
+            let root = Hash256::from_slice(&root);
```

Comment:
Any chance we can use different errors here? Otherwise if something bad happens there is ambiguity.